contracts/utils/Bytes.sol

```diff
@@ -0,0 +1,18 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+
+import {Math} from "./math/Math.sol";
+
+/**
+ * @dev Bytes operations.
+ */
+library Bytes {
+    /// @dev Reads a bytes32 from a bytes array without bounds checking.
+    function unsafeReadBytesOffset(bytes memory buffer, uint256 offset) internal pure returns (bytes32 value) {
+        // This is not memory safe in the general case
```

Comment:
Why is it not memory safe if it's just reading? I think we can safely annotate it:

Reply:
this is not memory safe because it's possible that the block reads from memory outside the bounds allocated by solidity (it's not checking bounds)

Reply:
Ah got it. I consider it may be better to annotate as `memory-safe` and check bounds. Otherwise, [optimizations get disabled](https://forum.soliditylang.org/t/assembly-and-solidity-optimizations/2053/3), perhaps it's cheaper to check and annotate than not checking (when optimizations are applied)

Reply:
The entier point of this function is to not check the bound. Checking the bound require loading the length, which is one expensive mload.

Reply:
I thought the entire point was to make it cheaper. I think disabling optimizations we're doing the opposite.

Reply:
Basically this function is already doing stuff as simple as possible. Not sure any further optimization is required from the compiler for this block.

Reply:
From [here](https://forum.soliditylang.org/t/assembly-and-solidity-optimizations/2053/3):

> Assuming that you’re using the via-ir pipeline, having assembly blocks that are not memory safe (i.e. have memory access within the block, but are not explicitly marked as memory safe) will prevent the ‘stack-to-memory’ mover, and will also not inline as aggressively (inlining is reduced to smaller functions than would be the case with memory safe blocks)

My point is that by having these assembly block that's not memory safe we might be hurting the performance of the whole contract.

Reply:
so: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5166#discussion_r1765350468

Reply:
Although I agree on adding a `Bytes` library, I'm not sure this particular function should go in. Perhaps this whole discussion indicates that the function should remain private in its contract.

I'd be in favor of not adding the Bytes library right now

Reply:
If it was just in this contract I'd agree.

But the truth is that this function is needed:
- in Strings.sol
- in Governor.sol
- in the other Bytes operation that come in #5252

So if we keep it private, we have to define it 3 times, in 3 different contracts/libraries.